Job to Remove Datastore Tables for Unsupported Resource Types

JVickery-TBS opened the pull request:
feat(dev): implemented IDomainObjectModification;

- Implemented `IDomainObjectModification` instead of `IResourceUrlChange`.
- Queued new job callback for unsupported datastore resources.
- Added new config option `ckanext.xloader.limit_datastore_tables`.


See https://github.com/open-data/ckanext-canada/pull/1382 for more discussion.

JVickery-TBS committed:
feat(dev): implemented IDomainObjectModification;

- Implemented `IDomainObjectModification` instead of `IResourceUrlChange`.
- Queued new job callback for unsupported datastore resources.
- Added new config option `ckanext.xloader.limit_datastore_tables`.

JVickery-TBS committed:
feat (dev): move method outside of class, added plugin loaded check;

- Moved job callback outside of plugin class.
- Added check for `datastore` plugin loaded.

JVickery-TBS commented:
Just working on some better conditions to prevent multiple jobs queuing up

JVickery-TBS committed:
feat(dev): added more conditions;

- Added method for conditions to queue jobs.

JVickery-TBS commented:
@wardi This one is ready for review.

I modified the `IResourceUrlChange` implementation into `IDomainObjectModification` implementation.

The new method to check if a datastore table should be removed (`_should_remove_unsupported_resource_from_datastore`) is quite chunky unfortunately. This was because the entity in the `notify` method is the `ckan.model.Resource` object and not the `dict` like it is in the job callback.

Not sure how to make that slimmer? Would I be able to do something like `res = session.query(model.Resource).get(resource_id)`? A session object does not get passed to the `notify` method, but the class seems to implement some type of session thing: https://github.com/open-data/ckan/blob/canada-v2.8/ckan/model/modification.py

Would I be able to get the session from the toolkit?

wardi commented on ckanext/xloader/plugin.py:
```diff
@@ -208,3 +229,50 @@ def get_helpers(self):
+    if not toolkit.asbool(toolkit.config.get('ckanext.xloader.limit_datastore_tables', False)):
+        return False
+    if isinstance(res_dict_or_obj, Resource):
+        return ((not XLoaderFormats.is_it_an_xloader_format(getattr(res_dict_or_obj, 'format', u''))
+                    or getattr(res_dict_or_obj, 'url_changed', False))
+                and (getattr(res_dict_or_obj, 'url_type') == 'upload'
+                    or getattr(res_dict_or_obj, 'url_type') == '')
+                and (getattr(res_dict_or_obj, 'datastore_active', False) 
+                    or getattr(res_dict_or_obj, 'extras', {}).get('datastore_active', False)))
```
this is kind of a hack, but so is passing either a dict or Resource object to the same function.
```suggestion
        res_dict_or_obj = res_dict_or_obj.__dict__
```

wardi commented on ckanext/xloader/plugin.py:
```diff
@@ -208,3 +229,50 @@ def get_helpers(self):
+        log.error('Resource %s does not exist.' % res['id'])
+        return
+
+    # only remove datastore tables from dataset types
```
```suggestion
    # only remove datastore tables from dataset types (canada fork only)
```

wardi commented on ckanext/xloader/plugin.py:
```diff
@@ -208,3 +229,50 @@ def get_helpers(self):
+
+
+def _should_remove_unsupported_resource_from_datastore(res_dict_or_obj):
+    if not toolkit.asbool(toolkit.config.get('ckanext.xloader.limit_datastore_tables', False)):
```
A better config option name, remember to document it.

```suggestion
    if not toolkit.asbool(toolkit.config.get('ckanext.xloader.clean_datastore_tables', False)):
```

wardi commented on ckanext/xloader/plugin.py:
```diff
@@ -109,11 +113,28 @@ def configure(self, config_):
+            p.toolkit.enqueue_job(fn=_remove_unsupported_resource_from_datastore, args=[entity.id])
+
+        # disable automatic submission of resource to xloader
+        # if validation is enabled or the url has not changed
```
```suggestion
        # if validation is enabled or the url has not changed (canada fork only)
```

JVickery-TBS committed:
Update ckanext/xloader/plugin.py

Co-authored-by: Ian Ward <ian@excess.org>

JVickery-TBS committed:
Update ckanext/xloader/plugin.py

Co-authored-by: Ian Ward <ian@excess.org>

JVickery-TBS committed:
Update ckanext/xloader/plugin.py

Co-authored-by: Ian Ward <ian@excess.org>

JVickery-TBS committed:
Update ckanext/xloader/plugin.py

Co-authored-by: Ian Ward <ian@excess.org>

JVickery-TBS committed:
feat(dev): added to readme;

- Added `clean_datastore_tables` config option to readme.

JVickery-TBS committed:
feat(dev): removed github workflow tests, add requirement;

- Added `ckanapi` to extension requirements.
- Removed github workflow test file.

JVickery-TBS commented:
@wardi okay I think this should be good now. The new config option has been added to the readme (will go into the declaration yaml upstream of course too).

I removed the github test workflow for this fork now too.

wardi commented:
can we add a flag to disable tests instead of removing the whole files, to reduce future conflicts?

JVickery-TBS committed:
revert(dev): reverted workflow removal;

- Reverted removal of github workflow file.

JVickery-TBS committed:
feat(dev): set workflow to dispatch only;

- Set github workflow to dispatch.

JVickery-TBS commented:
@wardi workflow file brought back and just changed on action to dispatch.
